enclave-runtime/src/rpc/worker_api_direct.rs

```diff
@@ -295,69 +285,52 @@ where
 	io
 }
 
-fn side_chain_io_handler() -> IoHandler {
+pub fn side_chain_io_handler<ImportFn>(import_fn: ImportFn) -> IoHandler
+where
+	ImportFn: Fn(Vec<SignedBlock>) -> Result<(), crate::error::Error> + Sync + Send + 'static,
+{
 	let mut io = IoHandler::new();
 
 	let sidechain_import_import_name: &str = "sidechain_importBlock";
-	io.add_sync_method(sidechain_import_import_name, |sidechain_blocks: Params| {
-		// Todo: actually do something with the block, i.e., apply the state diff.
-		// However, this requires importing the stf here. So, before doing that and increase
-		// the size of init_io_handler even more, we should think about how we can do that more
-		// modularized.
+	io.add_sync_method(sidechain_import_import_name, move |sidechain_blocks: Params| {
 		debug!("sidechain_importBlock rpc. Params: {:?}", sidechain_blocks);
+
 		let block_vec: Vec<u8> = sidechain_blocks.parse()?;
+
 		let blocks: Vec<SignedBlock> = Decode::decode(&mut block_vec.as_slice()).map_err(|_| {
 			jsonrpc_core::error::Error::invalid_params_with_details(
 				"Could not decode Vec<SignedBlock>",
 				block_vec,
 			)
 		})?;
+
 		info!("sidechain_importBlock. Blocks: {:?}", blocks);
+
+		let _ = import_fn(blocks).map_err(|e| {
+			jsonrpc_core::error::Error::invalid_params_with_details("Failed to import Block.", e)
+		})?;
+
 		Ok(Value::String("ok".to_owned()))
 	});
 
 	io
 }
 
-/// this is reduced to the side chain block import RPC interface (i.e. worker-worker communication)
-/// the entire rest of the RPC server is run inside the enclave and does not use this e-call function anymore
-#[no_mangle]
-pub unsafe extern "C" fn call_rpc_methods(
-	request: *const u8,
-	request_len: u32,
-	response: *mut u8,
-	response_len: u32,
-) -> sgx_status_t {
-	//init
-	let io = side_chain_io_handler();
-	// get request string
-	let req: Vec<u8> = from_raw_parts(request, request_len as usize).to_vec();
-	let request_string = match str::from_utf8(&req) {
-		Ok(req) => req,
-		Err(e) => {
-			error!("Decoding Header failed. Error: {:?}", e);
-			return sgx_status_t::SGX_ERROR_UNEXPECTED
-		},
-	};
-	// Rpc Response String
-	let response_string = io.handle_request_sync(request_string).unwrap();
-
-	// update response outside of enclave
-	let response_slice = from_raw_parts_mut(response, response_len as usize);
-	write_slice_and_whitespace_pad(response_slice, response_string.as_bytes().to_vec());
-	sgx_status_t::SGX_SUCCESS
-}
-
```

Comment:
This is moved to the main `lib.rs` file? In order to have all the e-call implementations in one place?

Reply:
Not exactly. I wanted to avoid having imports from `sidechain_impl.rs` in there.